plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,180 @@
+    @Nullable
+    protected E exporter;
+
+    /** The export interval of this exporter, default to 1 day */
+    protected TimeValue exportInterval = QueryInsightsSettings.MIN_EXPORT_INTERVAL;
+
+    /** The internal OpenSearch thread pool that execute async processing and exporting tasks*/
+    protected final ThreadPool threadPool;
+    protected volatile Scheduler.Cancellable scheduledFuture;
+
+    @Inject
+    public QueryInsightsService(ThreadPool threadPool, @Nullable S store, @Nullable E exporter) {
+        this.threadPool = threadPool;
+        this.store = store;
+        this.exporter = exporter;
+    }
+
+    /**
+     * Ingest one record to the query insight store
+     *
+     * @param record the record to ingest
+     */
+    protected void ingestQueryData(R record) {
+        if (this.store != null) {
+            this.store.add(record);
+        }
+    }
+
+    /**
+     * Get all records that are in the query insight store,
+     * By default, return the records in sorted order.
+     *
+     * @return List of the records that are in the query insight store
+     * @throws IllegalArgumentException if query insight is disabled in the cluster
+     */
+    public List<R> getQueryData() throws IllegalArgumentException {
+        if (!enableCollect) {
+            throw new IllegalArgumentException("Cannot get query data when query insight feature is not enabled.");
+        }
+        clearOutdatedData();
+        List<R> queries = new ArrayList<>(store);
+        queries.sort(Collections.reverseOrder());
+        return queries;
+    }
+
+    /**
+     * Clear all outdated data in the store
+     */
+    public abstract void clearOutdatedData();
+
+    /**
+     * Restart the exporter with new config
+     */
+    public abstract void resetExporter(boolean enabled, QueryInsightsExporterType type, String identifier);
+
+    /**
+     * Clear all data in the store
+     */
+    public void clearAllData() {
+        store.clear();
+    }
+
+    public void setEnableCollect(boolean enableCollect) {
+        this.enableCollect = enableCollect;
+    }
+
+    public boolean getEnableCollect() {
+        return this.enableCollect;
+    }
+
+    public void setEnableExport(boolean enableExport) {
+        this.enableExport = enableExport;
+    }
+
+    public boolean getEnableExport() {
+        return this.enableExport;
+    }
+
+    /**
+     * Start the Query Insight Service.
+     */
+    @Override
+    protected void doStart() {
+        if (exporter != null && getEnableExport()) {
+            scheduledFuture = threadPool.scheduleWithFixedDelay(this::doExport, exportInterval, ThreadPool.Names.GENERIC);
+        }
+    }
+
+    /**
+     * Stop the Query Insight Service
+     */
+    @Override
+    protected void doStop() {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel();
+            if (exporter != null && getEnableExport()) {
+                doExport();
```

Comment:
Is there reference service within OpenSearch that does similar export while stopping? I just need to understand the impact of sync vs async approach

Reply:
I don't think we have any other similar examples. What are the concerns here to export data when stopping the service?